OSDI22/microbench/test_pipeline.py

```diff
@@ -0,0 +1,113 @@
+import ray
+import csv
+import argparse
+import numpy as np 
+import time
+from time import perf_counter
+
+####################
+## Argument Parse ##
+####################
+parser = argparse.ArgumentParser()
+parser.add_argument('--WORKING_SET_RATIO', '-w', type=int, default=3)
+parser.add_argument('--OBJECT_STORE_SIZE', '-o', type=int, default=1_005_000_000)
+parser.add_argument('--OBJECT_SIZE', '-os', type=int, default=40_000_000)
+parser.add_argument('--RESULT_PATH', '-r', type=str, default="../data/dummy.csv")
+parser.add_argument('--NUM_STAGES', '-ns', type=int, default=1)
+parser.add_argument('--NUM_TRIAL', '-t', type=int, default=5)
+args = parser.parse_args()
+params = vars(args)
+
+OBJECT_STORE_SIZE = params['OBJECT_STORE_SIZE'] 
+OBJECT_SIZE = params['OBJECT_SIZE'] 
+WORKING_SET_RATIO = params['WORKING_SET_RATIO']
+RESULT_PATH = params['RESULT_PATH']
+NUM_STAGES = params['NUM_STAGES']
+NUM_TRIAL = params['NUM_TRIAL']
+OBJECT_STORE_BUFFER_SIZE = 5_000_000 #this value is to add some space in ObjS for nprand metadata and ray object metadata
+
+def test_ray_pipeline():
+    ray_pipeline_begin = perf_counter()
+
+    @ray.remote(num_cpus=1)
+    def last_consumer(obj_ref):
+        return True
+
+    @ray.remote(num_cpus=1)
+    def consumer(obj_ref):
+        return np.zeros(OBJECT_SIZE // 8)
+
+    @ray.remote(num_cpus=1) 
+    def producer(): 
+        #time.sleep(0.1)
+        return np.zeros(OBJECT_SIZE // 8)
+        
+    num_fill_object_store = (OBJECT_STORE_SIZE//OBJECT_SIZE)//NUM_STAGES
+    refs = [[] for _ in range(NUM_STAGES+1)]
+    for _ in range(WORKING_SET_RATIO*num_fill_object_store):
+        refs[0].append(producer.remote())
+
+    for stage in range(1, NUM_STAGES):
+        for r in refs[stage-1]:
+            refs[stage].append(consumer.remote(r))
+
+    for r in refs[NUM_STAGES-1]:
+        refs[NUM_STAGES].append(last_consumer.remote(r))
+
+    #del produced_objs
```

Comment:
I think we need something like this line to remove the refs to the produced objects.